### PR TITLE
[NestedTensor] Use maybe_mark_dynamic instead of mark_dynamic

### LINKED ISSUE #127097: [NestedTensor] RelaxedUnspecConstraint failures due to mark_dynamic in NT constructor
### 🐛 Describe the bug

If a tensor is marked as `mark_dynamic`, then there is some logic later on to verify that the shape does in fact remain dynamic:

https://github.com/pytorch/pytorch/blob/87f79af24dae87712996c9f135f2912a95bdd029/torch/fx/experimental/symbolic_shapes.py#L3653-L3664

NJT constructor marks its values tensor as dynamic:

https://github.com/pytorch/pytorch/blob/87f79af24dae87712996c9f135f2912a95bdd029/torch/nested/_internal/nested_tensor.py#L122

This can cause issues, demonstrated in the repro below:

```python
import torch

values = torch.randn((18, 16))
offsets = torch.tensor([0, 2, 3, 6, 15, 18])
like_values = torch.randn_like(values)

# this marks values as dynamic
nt = torch.nested.nested_tensor_from_jagged(values, offsets)

def fn(values, same_size):
    return values + same_size

torch.compile(fn)(values, like_values)
```

If we do some operation on `values` which compares it to a tensor with static but otherwise equal shape, then we can specialize the dynamic shape of `values` and cause this check to fail

See https://github.com/pytorch/pytorch/issues/102814

### Versions

main branch of pytorch as of may 23, CPU-only build

cc @cpuhrsch @jbschlosser @bhosmer @drisspg @soulitzer

### PR DESCRIPTION
**Pull request by davidberard98**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #127453

Fixes #127097

**TL;DR**: dimensions marked with mark_dynamic can result in assertion failures if the marked-dynamic dimensions get specialized. In NJT, we don't care _that_ much that a dimension is marked as dynamic. So instead, mark with `maybe_mark_dynamic` which suggests that a dimension should be dynamic, but doesn't fail if the dimension gets specialized.

**Background**:
NJT marks the values tensor as dynamic:

https://github.com/pytorch/pytorch/blob/49ad90349d57c35ab83f40c28d8b18caefb416d1/torch/nested/_internal/nested_tensor.py#L122

It does this for two reasons:
1. **Conceptual**: We know that this dimension _should_ be dynamic; it's a nested tensor, so the sequence lengths will _probably_ vary between batches in the common case. Therefore, we should compile it as dynamic to prevent needing a recompile to trigger automatic dynamic shapes.
2. **Implementation detail**: Right now we run into issues with torch.compile / tensor_unflatten / other details when the dimensions are not marked as dynamic. We have some attempts to remove this (e.g. https://github.com/pytorch/pytorch/pull/126563) but while testing this I wasn't able to get all tests to pass, so there could be potential regressions here if we removed the mark_dynamic.

**Justification for this change**

1. **Conceptual**: AFAIK, we don't care enough about the dynamism of this dimension to error out if we specialize. We'd prefer that we don't have to recompile to get automatic dynamic shapes, but it's also better to not have this issue (and not to force the user to go hunt down all the other equivalent shapes to mark them as dynamic as well). This solution allows us to suggest the dynamism but not force it.
2. **Implementation detail**: This still marks the dimension as symbolic at the beginning of dynamo tracing, so we will (probably) avoid a lot of the issues we run into when we completely remove the `mark_dynamic` decorators.

Differential Revision: [D57933779](https://our.internmc.facebook.com/intern/diff/D57933779)

### CODE REVIEW

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D57933779).

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D57933779).

File: test/test_nestedtensor.py [hunk 1]
```diff
@@ -4008,6 +4009,70 @@ def check_size(nt1, nt2, nt3, nt4):
...
+            return values + same_size
+
+        compile_counter = torch._dynamo.testing.CompileCounter()
+
+        compiled_fn = torch._dynamo.optimize(compile_counter, nopython=True)(fn)
+        check_results(fn, compiled_fn, generate_inp(18))
+        self.assertEqual(compile_counter.frame_count, 1)
+
+        check_results(fn, compiled_fn, generate_inp(19))
+        # we'll probably recompile here with dynamic shapes - it's okay if not though.
```
Comment (soulitzer) [inline test/test_nestedtensor.py:4067]: hmm is the case where we don't recompile is just when the inputs happen to be the same size, and for this test, we expect to always recompile?

**Review by soulitzer — APPROVED** [review]
Pretty cool, didn't know maybe_mark_dynamic existed!

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D57933779).

**Review by YuqingJ — APPROVED** [review]

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D57933779).
